plugins/poll/lib/PollActions.php

```diff
@@ -0,0 +1,256 @@
+ *
+ * Package and location is not indicated
+ * Should not include any kaltura dependency in this class - to enable it to run in cache only mode
+ */
+class PollActions
+{
+
+	const ID_SEPARATOR_CHAR = '-';
+	const ANSWER_SEPARATOR_CHAR = ',';
+	const ID_NUM_ELEMENTS = 3;
+
+	/**
+	 * @var PollCacheHandler
+	 */
+	private static $pollsCacheHandler = null;
+	/**
+	 * @var array of const strings representing poll types
+	 */
+	private static $pollTypes = null;
+	/**
+	 * @var string
+	 */
+	private static $kalturaSecret = null;
+	/**
+	 * @var int
+	 */
+	private static $pollCacheTTL = 0;
+
+	/* Configuration */
+	/**
+	 *
+	 * @throws Exception
+	 */
+	private static function init()
+	{
+		$pollConf = kConf::get("poll");
+		if (array_key_exists('secret' ,$pollConf))
+			self::$kalturaSecret = $pollConf['secret'];
+		if (!self::$kalturaSecret)
+			throw new Exception("Could not find polls_secret in the configuration");
+		if (array_key_exists('cache_ttl', $pollConf))
+			self::$pollCacheTTL = $pollConf['cache_ttl'];
+		self::$pollsCacheHandler = new PollCacheHandler(self::$pollCacheTTL);
+		self::$pollTypes = array(
+			'SINGLE_ANONYMOUS',
+			'SINGLE_RESTRICT',
+			'MULTI_ANONYMOUS',
+			'MULTI_RESTRICT');
+	}
+
+	/* Poll Id Action */
+	public static function generatePollId($type = 'SINGLE_ANONYMOUS')
+	{
+		self::init();
+		if (!self::isValidPollType($type))
+			throw new Exception("Poll type provided is invalid");
+		$randKey = rand();
+		$hash = hash_hmac('SHA256', self::$kalturaSecret, $randKey);
+		return $type.self::ID_SEPARATOR_CHAR.$hash.self::ID_SEPARATOR_CHAR.$randKey;
+	}
+
+	private static function  isValidPollType($type)
+	{
+		return in_array($type, self::$pollTypes);
+	}
+
+	private static function isValidPollIdStructure($id)
+	{
+		self::init();
+		$idElements = explode(self::ID_SEPARATOR_CHAR, $id);
+		if (count($idElements) === self::ID_NUM_ELEMENTS ) {
+			$pollType = $idElements[0];
+			$hash = $idElements[1];
+			$key = $idElements[2];
+			$simulatedHash = hash_hmac('SHA256', self::$kalturaSecret, $key);
+			$isHashOk = strcmp($hash, $simulatedHash) === 0;
+			$validPollType = self::isValidPollType($pollType);
+			return $isHashOk && $validPollType;
+
+		}
+		return false;
+	}
+
+	/* Poll Vote Actions */
+	public static function vote($params)
+	{
+		$pollId = $params['pollId'];
+		$userId = $params['userId'];
+		$ansIds = $params['answerIds'];
+		return self::setVote($pollId, $userId, $ansIds);
+	}
+
+	public static function setVote($pollId, $userId, $ansIds)
+	{
+		if (!$pollId || !$userId || !$ansIds)
+			throw new Exception('Missing parameter for vote action');
+		$answers = explode(self::ANSWER_SEPARATOR_CHAR, $ansIds);
+		if (self::isValidPollIdStructure($pollId)) {
+			// check early user vote
+			$previousAnswers = self::$pollsCacheHandler->setCacheVote($userId, $pollId, $answers);
+			if ($previousAnswers) {
+				self::$pollsCacheHandler->decrementAnsCounter($pollId, $previousAnswers);
+			} else {

```

Comment:
if you vote with multi option one voters votes for multi answers which breaks that logic